util/rateLimit.go

```diff
@@ -0,0 +1,501 @@
+		runLock:                &deadlock.Mutex{},
+		window:                 d,
+		consumed:               make(chan event, bsize),
+		served:                 make(chan event, bsize),
+		shouldDropQueries:      make(chan shouldDropQuery, bsize),
+		targetRateRefreshTicks: bsize / 10, // have the Congestion Manager refresh its target rates every 10% through the queue
+		consumedByClient:       map[ErlClient]*[]time.Time{},
+		exp:                    4,
+		wg:                     sync.WaitGroup{},
+	}
+	return &ret
+}
+
+// Consumed implements CongestionManager by putting an event on the consumed channel,
+// to be processed by the Start() loop
+func (cm *redCongestionManager) Consumed(c ErlClient, t time.Time) {
+	select {
+	case cm.consumed <- event{
+		c: c,
+		t: t,
+	}:
+	default:
+	}
+}
+
+// Served implements CongestionManager by putting an event on the done channel,
+// to be processed by the Start() loop
+func (cm *redCongestionManager) Served(t time.Time) {
+	select {
+	case cm.served <- event{
+		t: t,
+	}:
+	default:
+	}
+}
+
+// ShouldDrop implements CongestionManager by putting a query shouldDropQueries channel,
+// and blocks on the response to return synchronously to the caller
+// if an error should prevent the query from running, the result is defaulted to false
+func (cm *redCongestionManager) ShouldDrop(c ErlClient) bool {
+	ret := make(chan bool)
+	select {
+	case cm.shouldDropQueries <- shouldDropQuery{
+		c:   c,
+		ret: ret,
+	}:
+		return <-ret
+	default:
+		return false
+	}
+}
+
+// Start will kick off a goroutine to consume activity from the different activity channels,
+// as well as service queries about if a given capacityQueue should drop
+func (cm *redCongestionManager) Start() {
+	// check if the maintainer is already running to ensure there is only one routine
+	cm.runLock.Lock()
+	defer cm.runLock.Unlock()
+	if cm.running {
+		return
+	}
+	cm.ctx, cm.ctxCancel = context.WithCancel(context.Background())
+	cm.running = true
+	cm.wg.Add(1)
+	go cm.run()
+}
+
+func (cm *redCongestionManager) run() {
+	tick := 0
+	targetRate := float64(0)
+	consumedByClient := map[ErlClient]*[]time.Time{}
+	serves := []time.Time{}
+	exit := false
+	for {
+		select {
+		// prioritize shouldDropQueries
+		case query := <-cm.shouldDropQueries:
+			cutoff := time.Now().Add(-1 * cm.window)
+			prune(consumedByClient[query.c], cutoff)
+			query.ret <- cm.shouldDrop(targetRate, query.c, consumedByClient[query.c])
+		default:
+			select {
+			// "should drop" queries
+			case query := <-cm.shouldDropQueries:
+				cutoff := time.Now().Add(-1 * cm.window)
+				prune(consumedByClient[query.c], cutoff)
+				query.ret <- cm.shouldDrop(targetRate, query.c, consumedByClient[query.c])
+			// consumed events -- a client has consumed capacity from a queue
+			case e := <-cm.consumed:
+				if consumedByClient[e.c] == nil {
+					ts := []time.Time{}
+					consumedByClient[e.c] = &ts
+				}
+				*(consumedByClient[e.c]) = append(*(consumedByClient[e.c]), e.t)
```

Comment:
what is the upper bound for this time stamp arrays?

Reply:
Strictly, no upper bound, but the length of these lists will be proportional to the service rate:
* client submits `backlogSize` transactions, filling the backlog
* txHandler is magic and can instantly dequeue all requests
* client fills the backlog again, repeat

However many times this can happen within the rolling window (10s) is how long the list could get, theoretically. Same answer for all-clients, since they'd share the ability to fill the backlog.

And obviously, window size is directly proportional to this as well, since a 20min window affords twice as much time to load the backlog.